MayssaraOmar/assignment2/Assignment 2.cpp

```diff
@@ -0,0 +1,275 @@
+	tree_node<T>* insert_right(tree_node<T>*, const T &); // insert an element to the right of the node // if an element to the right already exists, it overwrites it
+	tree_node<T>* get_root();
+	tree_node<T>* get_left(tree_node<T>*);
+	tree_node<T>* get_right(tree_node<T>*);
+	void remove_leaf(tree_node<T>*); // removes a leaf node // if the node isn't a leaf it doesn't remove anything
+	bool empty();
+	void clear();
+	void print_ancestors(const T &target); // print ancestors' keys starting from the parent to the root separated by spaces
+	pair<tree_node<T>*, T> common_ancestor(tree_node<T>*, tree_node<T>*); // in case of an error, the first part of the returned pair will be nullptr
+private:
+	bool print_ancestors_helper(tree_node<T>* current, const T &target);
+	pair<tree_node<T>*, int> common_ancestor_helper(tree_node<T>*,
+			tree_node<T>*, tree_node<T>*);
+
+};
+template<class T>
+binary_tree<T>::binary_tree() {
+	root = nullptr;
+}
+
+template<class T>
+binary_tree<T>::~binary_tree() {
+	clear();
+}
+
+template<class T>
+tree_node<T>* binary_tree<T>::insert_root(const T &data) {
+	return root = new tree_node<T>(data, nullptr, nullptr);
+}
+
+template<class T>
+tree_node<T>* binary_tree<T>::insert_left(tree_node<T>* current,
+		const T &data) {
+	if (current) {
+		if (!current->left) {
+			current->left = new tree_node<T>(data, nullptr, nullptr);
+		} else {
+			current->left->data = data;
+		}
+		return current->left;
+	}
+	return nullptr;
+}
+template<class T>
+tree_node<T>* binary_tree<T>::insert_right(tree_node<T>* current,
+		const T &data) {
+	if (current) {
+		if (!current->right) {
+			current->right = new tree_node<T>(data, nullptr, nullptr);
+
+		} else {
+			current->right->data = data;
+		}
+		return current->right;
+	}
+	return nullptr;
+}
+template<class T>
+tree_node<T>* binary_tree<T>::get_root() {
+	return root;
+}
+template<class T>
+tree_node<T>* binary_tree<T>::get_left(tree_node<T>* current) {
+	if (current) {
+		return current->left;
+	}
+	return nullptr;
+}
+template<class T>
+tree_node<T>* binary_tree<T>::get_right(tree_node<T>* current) {
+	if (current) {
+		return current->right;
+	}
+	return nullptr;
+}
+template<class T>
+void binary_tree<T>::remove_leaf(tree_node<T>* current) {
+	if (current && !current->left && !current->right) {
+		delete current;
+	}
+}
+
+template<class T>
+bool binary_tree<T>::empty() {
+	return root == nullptr;
+}
+template<class T>
+void binary_tree<T>::clear() {
+	if (empty()) {
+		return;
+	}
+	stack<tree_node<T>*> st;
+	st.push(root);
+	while (!st.empty()) {
+		tree_node<T>* node = st.top();
+		st.pop();
+		if (node->right)
+			st.push(node->right);
+		if (node->left)
+			st.push(node->left);
+		delete node;
+	}
+	root = nullptr;
+	return;
+}
+template<class T>
+bool binary_tree<T>::print_ancestors_helper(tree_node<T>* current,
+		const T &target) {
+	if (!current) {
+		return false;
+	}
+
+	if (print_ancestors_helper(current->left, target)) {
+		cout << current->data << ' ';
```

Comment:
Please, make the function return the list of ancestors (in a `vector<T>` for example) and print it on main.